lua: adds dynamicTypedMetadata() on streamInfo()

agrawroh opened the pull request:
## Description

This PR adds support for accessing `typed_filter_metadata` through the stream info API. This allows Lua scripts to safely access protobuf-encoded metadata, with proper unpacking of **Any** messages into Lua tables. The implementation just extends the support for `dynamicTypedMetadata()` which we added for `connectionStreamInfo()` in the earlier PR.

---

**Commit Message:** lua: adds `dynamicTypedMetadata()` on `streamInfo()`
**Additional Description:** Added a new ``dynamicTypedMetadata()`` on ``streamInfo()`` which could be used to access the typed metadata from HTTP filters, such as the ExtProc, etc.
**Risk Level:** Low
**Testing:** Unit tests and integration tests added
**Docs Changes:** Added
**Release Notes:** Added

wbpcode commented on source/extensions/filters/http/lua/wrappers.cc:
```diff
@@ -172,6 +173,64 @@ int StreamInfoWrapper::luaDynamicMetadata(lua_State* state) {
   return 1;
 }
 
+int StreamInfoWrapper::luaDynamicTypedMetadata(lua_State* state) {
+  // Get filter name from Lua argument
+  const absl::string_view filter_name = Filters::Common::Lua::getStringViewFromLuaString(state, 2);
+
+  // Get the typed metadata from the stream's metadata
+  const auto& typed_metadata = stream_info_.dynamicMetadata().typed_filter_metadata();
+  const auto it = typed_metadata.find(filter_name);
+
+  if (it == typed_metadata.end()) {
+    // Return nil if the filter name is not found
+    lua_pushnil(state);
+    return 1;
+  }
+
+  // The typed metadata is stored as a ProtobufWkt::Any
+  const ProtobufWkt::Any& any_message = it->second;
+
+  // Extract the type name from the type URL
+  const std::string& type_url = any_message.type_url();
+  const size_t pos = type_url.find_last_of('/');
+  if (pos == std::string::npos || pos >= type_url.length() - 1) {
+    ENVOY_LOG(debug, "Invalid type URL in typed metadata for filter {}: {}", filter_name, type_url);
+    lua_pushnil(state);
+    return 1;
+  }
+  const absl::string_view type_name = absl::string_view(type_url).substr(pos + 1);
+
+  // Get the descriptor pool to find the message type
+  const auto* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(std::string(type_name));
+
+  if (descriptor == nullptr) {
+    ENVOY_LOG(debug, "Cannot find descriptor for type: {}", type_name);
+    lua_pushnil(state);
+    return 1;
```
Couldn't this share same function with the connection level dynamic typed metadata?

agrawroh commented:
cc @ggreenway Gentle ping on this

agrawroh commented:
> Any way to add some tests to source/extensions/filters/common/lua in order to not lose coverage?

As I moved the method from `source/extensions/filters/http/lua/` => `source/extensions/filters/common/lua/`. The numbers got moved. I added some test cases to improve it but there is one small case which I'm not able to cover and hence the slight drop of `0.1`.

RyanTheOptimist commented:
> > Any way to add some tests to source/extensions/filters/common/lua in order to not lose coverage?
> 
> As I moved the method from `source/extensions/filters/http/lua/` => `source/extensions/filters/common/lua/`. The numbers got moved. I added some test cases to improve it but there is one small case which I'm not able to cover and hence the slight drop of `0.1`.

Fair enough. coverage LGTM
